presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaSplitManager.java

```diff
@@ -93,44 +87,38 @@ public ConnectorSplitSource getSplits(
-            TopicMetadataResponse topicMetadataResponse = simpleConsumer.send(topicMetadataRequest);
+            KafkaConsumer<String, String> consumer = consumerManager.createConsumer(Thread.currentThread().getId(), node);
+            List<PartitionInfo> parts = consumer.partitionsFor(topic);
+            consumer.close();
 
             ImmutableList.Builder<ConnectorSplit> splits = ImmutableList.builder();
 
-            for (TopicMetadata metadata : topicMetadataResponse.topicsMetadata()) {
-                for (PartitionMetadata part : metadata.partitionsMetadata()) {
-                    log.debug("Adding Partition %s/%s", metadata.topic(), part.partitionId());
-
-                    BrokerEndPoint leader = part.leader();
-                    if (leader == null) {
-                        throw new PrestoException(GENERIC_INTERNAL_ERROR, format("Leader election in progress for Kafka topic '%s' partition %s", metadata.topic(), part.partitionId()));
-                    }
-
-                    HostAddress partitionLeader = HostAddress.fromParts(leader.host(), leader.port());
-
-                    SimpleConsumer leaderConsumer = consumerManager.getConsumer(partitionLeader);
-                    // Kafka contains a reverse list of "end - start" pairs for the splits
-
-                    long[] offsets = findAllOffsets(leaderConsumer, metadata.topic(), part.partitionId());
-
-                    for (int i = offsets.length - 1; i > 0; i--) {
-                        KafkaSplit split = new KafkaSplit(
-                                connectorId,
-                                metadata.topic(),
-                                kafkaTableHandle.getKeyDataFormat(),
-                                kafkaTableHandle.getMessageDataFormat(),
-                                kafkaTableHandle.getKeyDataSchemaLocation().map(KafkaSplitManager::readSchema),
-                                kafkaTableHandle.getMessageDataSchemaLocation().map(KafkaSplitManager::readSchema),
-                                part.partitionId(),
-                                offsets[i],
-                                offsets[i - 1],
-                                partitionLeader);
-                        splits.add(split);
-                    }
+            for (PartitionInfo part : parts) {
+                Node leader = part.leader();
+                if (leader == null) {
+                    throw new PrestoException(GENERIC_INTERNAL_ERROR, format("Leader election in progress for Kafka topic '%s' partition %s", topic, part.partition()));
                 }
+
+                HostAddress partitionLeader = HostAddress.fromParts(leader.host(), leader.port());
+                KafkaTableLayoutHandle layoutHandle = (KafkaTableLayoutHandle) layout;
+                long startTs = layoutHandle.getOffsetStartTs();
+                long endTs = layoutHandle.getOffsetEndTs();
+
+                KafkaSplit split = new KafkaSplit(
```

Comment:
could you please elaborate more on why pushdown offsets computation for split?
could we compute offsets here for all splits, if the computation overhead is the same? As Coordinator decouples split generation with split scan

Reply:
Before pushdown, offset computation is done split by split in serial in a big for loop, by a single thread I believe. 
If we delay offset computation to after splits are distributed to workers, this computation can be done in parallel by multiple threads on multiple workers. 
Basically we gain performance benefit by parallelizing the work.

Reply:
how expensive is offset computation? Do you have benchmarks shows how much performance gain we get by pushdown offset computation?
The logic is a more complicated after pushdown. we could do the pushdown if get big performance gain.

Reply:
Here are some numbers. For  a topic with 1000 partitions, it takes average of 40120 millisecond to build all 1000 splits before the pushdown. It's done on a single machine (the coordinator).
After pushdown, this can be finished for 1 millisecond.
Above number are average from 10 runs. 

The gain is about a fixed 41 seconds. Whether it's worth it really depends on the use case, and cluster settings. For example, if we are interested in 1-minute data from Kafka, then we can reduce 1 minute execution time to couple seconds. But if we are interested in 1-hour data, then it's saving 40 seconds from several minutes. 

I will leave this decision to you.

Reply:
thank you, Da. I did some research, seems most Kafka topics has < 100 partitions, Linkedin has average ~70 partitions per topic: https://conferences.oreilly.com/strata/strata-ca-2018/public/schedule/detail/63921
Am inclined to the following:
at a first step, have an implementation without pushdown, as most users will not have hundreds of topic partitions, and the performance penalty is not big. 
We could file a following improvement PR, to discuss the details of pushdown implementation. (the current pushdown logic is kind of complicated)
what do you think?

Reply:
Sound good. I will update the PR to remove it.